### PR TITLE
feat(pfd): Add vertical deviation indicator, ROW/ROP/OANS warnings

### PR DESCRIPTION
**Pull request by flogross89**
## Summary of Changes
Relevant for both A380X and A32NX. 

- Adds OANS, ROP and ROW warnings to the PFD and aural alerts
- Transmits new EGPWS discrete words from the GPWS module
- Implements stall warning (aural alert and PFD warning)

A380X only:
- Adds GPWS alerts to PFD
- Ports the VDEV indicator, as well as the upper and lower speed limits from the a32nx to the a380x
- Adds LS reminder for ILS, subsequently move marker (OM/MM/IM) information
- ILS information: Frequency now has the same font size before and after decimal point, see references

## Screenshots (if necessary)
image
image
image
image
image
image
image
image
image
image

A380X only:
image
image
image
image
image
image
image
image

## References
https://www.youtube.com/watch?v=EorNGwNTzwQ
https://www.navblue.aero/product/rops-plus/
https://safetyfirst.airbus.com/app/themes/mh_newsdesk/documents/archives/the-runway-overrun-prevention-system.pdf
https://www.youtube.com/watch?v=FbCIjWaMqTI
https://www.youtube.com/watch?v=hDh_9MEz-hg
image

## Additional context

Discord username (if different from GitHub): floridude

## Testing instructions 
Testing of ROW/ROP on A32NX not yet possible. 

**For A32NX:**
**Test 1: Stall warning**
Bring aircraft into alternate law (e.g. by turning both SECs to OFF), throttle to idle and maximum pitch up, until the aircraft enters a stall.
An aural warning "STALL, STALL", shall be heard.

**Test 2: GPWS warnings**
Please perform a flight triggering the following GPWS warnings, and confirm that there has been no regression, i.e. they still trigger:

- Sinkrate
- Pull Up
- Terrain
- Don't Sink
- Too Low Gear
- Too Low Flaps
- Too Low Terrain
- Glideslope

**Test 3: ROP/ROW Warnings in A32NX**
To trigger these manually, set the local variable A32NX_ROW_ROP_WORD_1 to the following value: 
MAX BRAKING: 14046724096.000
MAX REVERSE: 14053015552.000
IF WET: RWY TOO SHORT: 14060355584.000
RWY TOO SHORT: 14068219904.000

## How to download the PR for QA

Every new commit to this PR will cause new A32NX and A380X artifacts to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on either **flybywire-aircraft-a320-neo** or **flybywire-aircraft-a380-842** download link at the bottom of the page

### CODE REVIEW

**Review by 2hwk — APPROVED** [review]

**Comment by Popespice** [comment]
Quality Assurance Trainee Report

Discord          : Popepsice
Object of testing: #8605 
Tier of Testing  : 2
Date             : 02042024

Testing Process: A32NX (Note, aircraft was on the ground during testing as feature is not ready for in-flight testing yet)

1. Log into sim
2. Load into aircraft
3. Power on through expedited power-on through flypad
4. Open Settings
5. Turn on developer mode
6. Navigate to top toolbar and follow this path: Tools --> Behaviors --> LocalVariables
7. Manually change the following from 0.0 to 1.0
"MAX REVERSE": A32NX_AUTOBRAKE_ROP_ACTIVE
"MAX BRAKING + MAX REVERSE": A32NX_MANUAL_BRAKING_ROP_ACTIVE
"IF WET:RWY TOO SHORT": A32NX_ROW_IF_WET_RWY_TOO_SHORT
"RWY TOO SHORT": A32NX_ROW_RWY_TOO_SHORT
"RWY AHEAD": A32NX_OANS_RWY_AHEAD
8. After each message displays on screen, change variable back from 1.0 to 0.0
9. Observe each message displays correctly

Negatives:       
N/A

Testing Results: 
Passed

Conclusions: 
Works as expected. Actions do not trigger in flight, all testing was performed on the ground.

Media:       
N/A

File: fbw-a32nx/src/systems/instruments/src/PFD/shared/BrakingWarningsPublisher.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+export class BrakingWarningsPublisher extends SimVarPublisher<BrakingWarningsSimVars> {
+    constructor(bus: EventBus) {
+        super(new Map([
+            ['autobrakeRopActive', { name: 'L:A32NX_AUTOBRAKE_ROP_ACTIVE', type: SimVarValueType.Bool }],
+            ['manualBrakingRopActive', { name: 'L:A32NX_MANUAL_BRAKING_ROP_ACTIVE', type: SimVarValueType.Bool }],
+            ['rowIfWetRwyTooShort', { name: 'L:A32NX_ROW_IF_WET_RWY_TOO_SHORT', type: SimVarValueType.Bool }],
+            ['rowRwyTooShort', { name: 'L:A32NX_ROW_RWY_TOO_SHORT', type: SimVarValueType.Bool }],
+            ['oansRwyAhead', { name: 'L:A32NX_OANS_RWY_AHEAD', type: SimVarValueType.Bool }],
```
Comment (tracernz) [inline fbw-a32nx/src/systems/instruments/src/PFD/shared/BrakingWarningsPublisher.ts:18]: I expect these should all come as bits in an ARINC word?

**Review by tracernz — CHANGES_REQUESTED** [review]
See dev channel.

**Comment by tracernz** [comment]
STALL STALL is a bit too high
image

**Comment by 2hwk** [comment]
Impacted by the potential future merge freeze. https://github.com/flybywiresim/aircraft/pull/8466 will be required to be merged into this PR at some point.

**Comment by flogross89** [comment]
lint changes completed

**Comment by 2hwk** [comment]
Noticing some weirdness with this branch when powering on the A/C, i.e. master caution/warning sounds which should not happen. 

Maybe related to pseudoFWC changes?

**Comment by flogross89** [comment]
> Noticing some weirdness with this branch when powering on the A/C, i.e. master caution/warning sounds which should not happen.
> 
> Maybe related to pseudoFWC changes?

Is the behavior correct on master, i.e. are we sure that this behavior is introduced with this PR? There's no change where I could explain this behavior from :/

**Comment by 2hwk** [comment]
> > Noticing some weirdness with this branch when powering on the A/C, i.e. master caution/warning sounds which should not happen.
> > Maybe related to pseudoFWC changes?
> 
> Is the behavior correct on master, i.e. are we sure that this behavior is introduced with this PR? There's no change where I could explain this behavior from :/

Definitely not the case in the current master branch, I will check again this weekend.

**Comment by flogross89** [comment]
> > > Noticing some weirdness with this branch when powering on the A/C, i.e. master caution/warning sounds which should not happen.
> > > Maybe related to pseudoFWC changes?
> > 
> > 
> > Is the behavior correct on master, i.e. are we sure that this behavior is introduced with this PR? There's no change where I could explain this behavior from :/
> 
> Definitely not the case in the current master branch, I will check again this weekend.

See ping in Discord, I'd be happy if you could somehow help me reproduce this :) If it's occuring in the A32NX, I don't know which sounds to expect at startup

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 2]
```diff
@@ -380,4 +381,27 @@ class A32NX_FWC {
...
+
+        // SET MAX REVERSE, if not already max. reverse set and !MAX_BRAKING
+        const maxReverseSet = SimVar.GetSimVarValue('L:XMLVAR_Throttle1Position', 'number') < 0.1 && SimVar.GetSimVarValue('L:XMLVAR_Throttle2Position', 'number') < 0.1;
+        const maxReverse = (w.getBitValueOr(12, false) || w.getBitValueOr(13, false)) && !maxReverseSet;
+        SimVar.SetSimVarValue('L:A32NX_AUDIO_ROW_SET_MAX_REVERSE', 'bool', !maxBraking && maxReverse);
+
+        // At 80kt, KEEP MAX REVERSE once, if max. reversers deployed
+        const ias = SimVar.GetSimVarValue('AIRSPEED INDICATED', 'knots');
+        SimVar.SetSimVarValue('L:A32NX_AUDIO_ROP_KEEP_MAX_REVERSE', 'bool', (ias <= 80 && ias > 4) && (w.getBitValueOr(12, false) || w.getBitValueOr(13, false)));
+
```
Comment (tracernz) [inline fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_FWC.js:405]: Hmm, we should not have any new stuff in here. We're trying to get rid of this file in favour of PseudoFWC (and eventual rust FWC).

File: fbw-a380x/src/base/flybywire-aircraft-a380-842/SimObjects/AirPlanes/FlyByWire_A380_842/panel/panel.cfg [hunk 3]
```diff
@@ -92,7 +92,7 @@ size_mm=512,512
 pixel_size=512,512
 texture=$ISIS2
 
-htmlgauge00=A32NX/ISIS/isis.html, 0,0,512,512
+;htmlgauge00=A32NX/ISIS/isis.html, 0,0,512,512
```
Comment (tracernz) [inline fbw-a380x/src/base/flybywire-aircraft-a380-842/SimObjects/AirPlanes/FlyByWire_A380_842/panel/panel.cfg:95]: Suggest doing ISIS changes in a different PR.
```suggestion
htmlgauge00=A32NX/ISIS/isis.html, 0,0,512,512
```

File: fbw-a380x/src/systems/instruments/src/Common/gauges.scss [hunk 4]
```diff
@@ -54,7 +54,7 @@
     }
 
     .GaugeThrustFill {
-        fill: $display-grey;
+        fill: #3A3A4B;
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/Common/gauges.scss:57]: We should use standard colours from our definitions for all EIS.

File: fbw-a380x/src/systems/instruments/src/PFD/style.scss [hunk 5]
```diff
@@ -244,19 +244,20 @@ text.Red {
...
-  fill: #0698ff;
+  fill: #0073DE;
 }
 .BlackFill {
   fill: $display-background;
 }
 
 .TapeBackground {
-  fill: $display-grey;
+  fill: #4A5071;
```
Comment (tracernz) [inline fbw-a380x/src/systems/instruments/src/PFD/style.scss:257]: As above, EIS colours should be all standard from our definitions. We don't want a mix of subtly different shades.

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_FWC.js (hunk 2 above)
Comment (flogross89) [reply fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_FWC.js:405]: Meaning I should put this in PseudoFWC?

File: fbw-a380x/src/systems/instruments/src/PFD/style.scss (hunk 5 above)
Comment (flogross89) [reply fbw-a380x/src/systems/instruments/src/PFD/style.scss:257]: These definitions are for the A32x though, the A380 used different colors (as determined by Mike and me). I will investigate the impact of changing the definition

Comment (tracernz) [reply fbw-a380x/src/systems/instruments/src/PFD/style.scss:257]: I can’t see any reason why all the elements in the A380 EIS should not use the exact same colours.

File: fbw-a380x/src/base/flybywire-aircraft-a380-842/SimObjects/AirPlanes/FlyByWire_A380_842/panel/panel.cfg (hunk 3 above)
Comment (flogross89) [reply fbw-a380x/src/base/flybywire-aircraft-a380-842/SimObjects/AirPlanes/FlyByWire_A380_842/panel/panel.cfg]: ok, moving to another PR

File: fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_FWC.js (hunk 2 above)
Comment (flogross89) [reply fbw-a32nx/src/base/flybywire-aircraft-a320-neo/html_ui/Pages/A32NX_Core/A32NX_FWC.js]: moved to PseudoFWC

**Comment by flogross89** [comment]
> > > Noticing some weirdness with this branch when powering on the A/C, i.e. master caution/warning sounds which should not happen.
> > > Maybe related to pseudoFWC changes?
> > 
> > 
> > Is the behavior correct on master, i.e. are we sure that this behavior is introduced with this PR? There's no change where I could explain this behavior from :/
> 
> Definitely not the case in the current master branch, I will check again this weekend.

This should be addressed now with the FWC timeout after startup

**Comment by noonebling** [comment]
Quality Assurance Trainee Report

Discord          : noonebling
Object of testing: 8605
Tier of Testing  : 2
Date             : 15/06/2024

Testing Process: 

Only for A32NX:
TEST 1:
Brought aircraft into alternate law, throttle idle and maximum pitch up, until the aircraft enters a stall
Aural warning "STALL, STALL" heard

TEST 2:
Performed flight triggering GPWS warnings for Sinkrate, Pull Up, Terrain, Don't Sink, Too Low Gear, Too Low Flaps, Too Low Terrain and Glideslope with no regression

TEST 3:
Set the local variable A32NX_ROW_ROP_WORD_1 to the following value:
MAX BRAKING: 14046724096.000
MAX REVERSE: 14053015552.000
IF WET: RWY TOO SHORT: 14060355584.000
RWY TOO SHORT: 14068219904.000
Checked for ROP warnings

Negatives:
N/A

Testing Results: 
Passed

Conclusions: 
Works as expected

**Comment by simaoaviation** [comment]
https://www.youtube.com/watch?v=s-pbRY7BIgk

alternate law, throttle idle, pitch up 50*, no stall warning.

**Pull request merged**